url4c41254592_topskychen_AuthTopk_tgz-pJ8-quickhull3d_QuickHull3DTestJ8/src/spatialindex/rtree/SecurityTree.java

```diff
@@ -0,0 +1,164 @@
+		System.out.println("build security info finished, time used:" + cpuTime + "ms");
+	}
+
+	public void buildSecurityTree(SecurityTree srtree, RTree _rtree) throws IndexOutOfBoundsException, Exception {
+		root = SecurityNode.buildSecurityNode(srtree, rtree);
+		level = root.getLevel();
+	}
+
+	public void saveSecurityNode(SecurityNode n, int id) {
+//		if (n.getLevel() > 1) {
+//			System.out.println("sNode[" + id + "] saved, lv=" + n.getLevel());
+//		}
+//		if(n.getLevel() > 1) 
+//			SecurityNodeMap.put(new Integer(id), n);
+	}
+
+	public SecurityNode getSecurityNode(int id) {
+		SecurityNode snode = SecurityNodeMap.get(new Integer(id));
+		if (snode == null) {
+			Long pos = SecurityNodeMapPos.get(new Integer(id));
+			if (pos == null) {
+				return null;
+			} else {
+				int length = SecurityNodeMapLength.get(new Integer(id)).intValue();
+				//load from file
+				try {
+					RandomAccessFile rfile = new RandomAccessFile(new File(filePath + ".sdat"), "r");
+					rfile.seek(pos.longValue());
+					byte[] data = new byte[length];
+					rfile.read(data);
+					rfile.close();
+//					if(id == 2){
+//						for(int i = 0; i < length; i++){
+//							System.out.print(data[i]);
+//						}System.out.println();
+//					}
+					snode = new SecurityNode(data, rtree);
+				} catch (FileNotFoundException e) {
+					e.printStackTrace();
+					return null;
+				} catch (IOException e) {
+					e.printStackTrace();
+					return null;
+				}
+				//cache it
+				saveSecurityNode(snode, id);
+				return snode;
+			}
+		} else {
+			return snode;
+		}
+	}
+
+	public String getRootEntaValue() {
+		return root.getEntaValue();
+	}
+
+	public String getrsarootentaValue(){
+		return root.getrsaentaValue();
+	}
+	
+	/**
+	 * auto add "idx" and "dat"
+	 * 
+	 * @param file
+	 * @throws IOException
+	 */
+	public static void writeToFile(DataOutputStream indexDos, DataOutputStream dataDos, int key, long dataPos, byte[] snodeData) throws IOException {
+		indexDos.writeInt(key);
+		indexDos.writeLong(dataPos);
+		if(dataPos < 0){
+			System.err.println("error of pas");
+		}
+		indexDos.writeInt(snodeData.length);
+		dataDos.write(snodeData);
+			//System.err.println("key : " + key + " pos : " + dataPos + " len : " + snodeData.length);
+	}
+	
+	public void clearcache(){
+		SecurityNodeMap.clear();
+	}
+	
+	public boolean checkindex(){
+		if(SecurityNodeMapLength.size() != SecurityNodeMapPos.size())return false;
+		Iterator<Integer> keyset = SecurityNodeMapPos.keySet().iterator();
+		while(keyset.hasNext()){
+			int key = keyset.next();
+			if(SecurityNodeMapPos.get(key) < 0){
+				System.err.println("error in pos");
+				return false;
+			}
+			if(SecurityNodeMapLength.get(key) < 0){
+				System.err.println("error in len");
+				return false;
+			}
+		}
+		return true;
+	}
+	
+	private void readFromFile(RTree rtree, String filePath) throws IOException {
+		DataInputStream indexDis = new DataInputStream(new FileInputStream(filePath + ".sidx"));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the following resource: *indexDis*. The resource is referenced at lines 146, 147, 148, 149, 151. The resource is closed at line 154. There are other execution paths that do not contain closure statements, for example, when FilterInputStream.available throws an exception. Ensure *indexDis* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).